### ISSUE TITLE
Comment display is incredibly slow in 1.2.0/1.2.0a/1.2.0b

### ISSUE THREAD

**Issue by bonesmoses** [talklittle/reddit-is-fun#164]
Since the 1.2.0 update, a bug fix has caused a comment loading regression. It is possible to watch the comments physically flow down the screen as if in slow motion, and the page loading icon continues to circle for up to several minutes. 1.1.4 definitely did not have this problem.

This is on an Android Eris running Gingerbread 2.3.5 after a reboot and Dalvik cache reset, though after reading a few app reviews, this is probably a problem in the app for all phones.

**Comment by talklittle** [comment]
Thanks for filing a bug for this. I'll improve the performance for the next release.

**Comment by sztomi** [comment]
I'd like to work on this and I'm currently profiling the application. Can you tell me which classes/methods are potential bottlenecks?

**Comment by talklittle** [comment]
The bottleneck entirely lives in the CommentsListActivity. problem is it is talking to the UI thread after processing each individual comment, even for comments that are far off screen. One approach may be to defer insertion of comments that are far off screen until the user scrolls closer to them.

Thanks for taking a look!

**Comment by sztomi** [comment]
Actually I think it's in DownloadCommentsTask.java (but I think you meant the same, that class used to live in the file you mentioned, according to the commit history). 
I observed that even though the comment loading stops (I can scroll to the bottom and nothing gets added) and the progressbar on the top disappears to signal this to the user, the spinner/throbber progress indicator still runs. What is supposed to stop it? Is is still loading the deferred comments?

**Comment by sztomi** [comment]
Adding each root node tree in one chunk improves the loading time by ~40% (that's me counting the seconds :P)
How much better was it before the regression? 

Edit: it's actually 2-3 times faster than the current upstream on larger threads (~1000 comments). Just one bug left and I need to understand the jumptarget stuff because I currently just commented that out. Is that documented?

**Comment by talklittle** [comment]
Before the regression it was much faster, maybe up to 10x?, considering it was only talking to the UI thread once, after having processed the comments entirely.

As for jump target, it's basically a strange priority system-- I go through the comment objects during or after (sorry I forget which; doesn't matter though) they've loaded from JSON, searching for the jump target. Jump target is simply the target comment to jump to e.g. when going to comment from inbox, or /r/bestof links maybe.

The really slow part of processing a comment is the Html.fromHtml() method--the main reason I made the big change to do individual comment inserts in the first place--so I defer the display of comment body for all comments before the jump target. The reason for deferring is they won't be visible anyway once you've jumped to the jump target. (You'd scroll up to see the deferred comments.) Once I've found the jump target, I proceed to do the fromHtml on deferred comments and comments appearing after the jump target.

Hope that helps. Agreed a lot of pieces of the code could do with documentation in the wiki... :)

**Comment by talklittle** [comment]
By the way thanks a lot for taking a look and I'm excited to hear about the huge performance gains!

**Comment by sztomi** [comment]
No problem, and thanks for the detailed answer. I'll keep on working on it when I have time and soon send pull request.

**Comment by talklittle** [comment]
sztomi, any news on this? If not, I'll work on a separate fix for the same slow-loading comments issue.

**Comment by sztomi** [comment]
Hi,

sorry for disappearing. Lots of work IRL :( I'll wrap it up in the weekend (if you can wait until that). If not, feel free to work on your fix.

**Comment by talklittle** [comment]
I think I'll be attempting a fix. Thanks for looking into it earlier.
I did cherry pick a change from one of your branches though--setting
the HTTP version to 1.1, which does seem to speed up
downloads/connections a bit (hope it's not just my imagination)!

On Tue, Nov 22, 2011 at 2:59 PM, sztomi
reply@reply.github.com
wrote:

**Comment by sztomi** [comment]
The main idea was to add list items in batch rather than one-by-one. According to the profiling that was the main problem because the layout was re-done for every item and that takes a looong time in long comment threads. I also wanted to remove the recursion and use a DFS, but I had a bug in that... that's not necessary though, the real bottleneck was the layout.

**Comment by talklittle** [comment]
Thanks for the info!

On Tue, Nov 22, 2011 at 4:30 PM, sztomi
reply@reply.github.com
wrote:

**Comment by talklittle** [comment]
Rather than taking the approach of inserting comments in chunks, I took a simple approach and divided comments cleanly into two phases:
1. Insertion of all comment list items in a single go. They use the plaintext comment body at first.
2. Do the "slow" comment processing in a separate AsyncTask, making use of onProgressUpdate to display the rendered HTML.

It's actually a perfect use case for a second AsyncTask, which keeps fields synchronized between a background thread and UI thread, but a lot easier than trying to work with the Thread primitive. When I tried using a Thread and calling runOnUiThread after each processed comment, the UI appeared to hang. But doing the same thing in AsyncTask worked like a charm.
